extensions/sinks/kafka/kafka.go

```diff
@@ -0,0 +1,221 @@
+
+func (m *kafkaSink) Configure(props map[string]interface{}) error {
+	c := &sinkConf{
+		Brokers:           "localhost:9092",
+		Topic:             "",
+		DeliveryGuarantee: AT_LEAST_ONCE,
+		SaslAuthType:      SASL_NONE,
+	}
+	if err := cast.MapToStruct(props, c); err != nil {
+		return err
+	}
+	if len(strings.Split(c.Brokers, ",")) == 0 {
+		return fmt.Errorf("brokers can not be empty")
+	}
+	if c.Topic == "" {
+		return fmt.Errorf("topic can not be empty")
+	}
+	if !(c.DeliveryGuarantee == AT_MOST_ONCE || c.DeliveryGuarantee == AT_LEAST_ONCE) {
+		return fmt.Errorf("deliveryGuarantee incorrect")
+	}
+	if !(c.SaslAuthType == SASL_NONE || c.SaslAuthType == SASL_SCRAM || c.SaslAuthType == SASL_PLAIN) {
+		return fmt.Errorf("saslAuthType incorrect")
+	}
+	if (c.SaslAuthType == SASL_SCRAM || c.SaslAuthType == SASL_PLAIN) && (c.SaslUserName == "" || c.SaslPassword == "") {
+		return fmt.Errorf("username and password can not be empty")
+	}
+
+	m.c = c
+	return nil
+}
+
+func (m *kafkaSink) Open(ctx api.StreamContext) error {
+	ctx.GetLogger().Debug("Opening kafka sink")
+	var maxAttempts int
+	var err error
+	var mechanism sasl.Mechanism
+	//delivery guarantee
+	switch m.c.DeliveryGuarantee {
+	case AT_MOST_ONCE:
+		maxAttempts = 0
+	case AT_LEAST_ONCE:
+		maxAttempts = 10
+	default:
+		maxAttempts = 10
+	}
+	//sasl authentication type
+	switch m.c.SaslAuthType {
+	case SASL_PLAIN:
+		mechanism = plain.Mechanism{
+			Username: m.c.SaslUserName,
+			Password: m.c.SaslPassword,
+		}
+	case SASL_SCRAM:
+		mechanism, err = scram.Mechanism(scram.SHA512, "username", "password")
+		if err != nil {
+			return err
+		}
+	default:
+		mechanism = nil
+	}
+	brokers := strings.Split(m.c.Brokers, ",")
+	w := &kafkago.Writer{
+		Addr:                   kafkago.TCP(brokers...),
+		Topic:                  m.c.Topic,
+		Balancer:               &kafkago.LeastBytes{},
+		Async:                  false,
+		AllowAutoTopicCreation: true,
+		MaxAttempts:            maxAttempts,
+		RequiredAcks:           -1,
+		Transport: &kafkago.Transport{
+			SASL: mechanism,
+		},
+		WriteTimeout: 10 * time.Second,
+		ReadTimeout:  10 * time.Second,
+	}
+	m.writer = w
+	return nil
+}
+
+func (m *kafkaSink) Collect(ctx api.StreamContext, item interface{}) error {
+	logger := ctx.GetLogger()
+	logger.Debugf("kafka sink receive %s", item)
+	var decodedBytes []byte
+	var err error
+	decodedBytes, _, err = ctx.TransformOutput(item)
+	if err != nil {
+		return fmt.Errorf("kafka sink transform data error: %v", err)
+	}
+
+	vctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	err = m.kafkaWriteWithBackoff(vctx, ctx.GetLogger().Errorf, 100*time.Millisecond, time.Second, kafkago.Message{Value: decodedBytes})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *kafkaSink) kafkaWriteWithBackoff(ctx context.Context, log func(string, ...interface{}), interval, maxInterval time.Duration, messages ...kafkago.Message) error {
+	var berr error
+	tries := 0
+
+retry:
```

Comment:
In previous kafkago.Writer initialization, there are properties like `MaxAttempts`, `WriteTimeout`. It looks like you are implementing retry and timeout mechanism in this function. Why do you need this if the Writer already has that?

Reply:
according to the official doc of segmentio/kafka-go , there is some extra error like LeaderNotAvailable which is not related to message sending , can be handle by this kind of code

https://github.com/segmentio/kafka-go
```
var err error
const retries = 3
for i := 0; i < retries; i++ {
    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
    defer cancel()
    
    // attempt to create topic prior to publishing the message
    err = w.WriteMessages(ctx, messages...)
    if errors.Is(err, LeaderNotAvailable) || errors.Is(err, context.DeadlineExceeded) {
        time.Sleep(time.Millisecond * 250)
        continue
    }

    if err != nil {
        log.Fatalf("unexpected error %v", err)
    }
}

if err := w.Close(); err != nil {
    log.Fatal("failed to close writer:", err)
}
```